src/codegen/function_body_chunk.rs

```diff
@@ -431,3 +511,75 @@ impl Builder {
             .next()
     }
 }
+
+fn c_type_mem_mode(env: &Env, parameter: &AnalysisCParameter) -> OutMemMode {
+    use self::OutMemMode::*;
+    match ConversionType::of(env, parameter.typ) {
+        ConversionType::Pointer => if parameter.caller_allocates {
+            UninitializedNamed(rust_type(env, parameter.typ).unwrap())
+        } else {
+            use library::Type::*;
+            let type_ = env.library.type_(parameter.typ);
+            match *type_ {
+                Fundamental(fund)
+                    if fund == library::Fundamental::Utf8
+                        || fund == library::Fundamental::Filename =>
+                        {
+                            if parameter.transfer == library::Transfer::Full {
+                                NullMutPtr
+                            } else {
+                                NullPtr
+                            }
+                        }
+                _ => NullMutPtr,
+            }
+        },
+        _ => Uninitialized,
+    }
+}
+
+fn type_mem_mode(env: &Env, parameter: &library::Parameter) -> OutMemMode {
+    use self::OutMemMode::*;
+    match ConversionType::of(env, parameter.typ) {
+        ConversionType::Pointer => if parameter.caller_allocates {
+            UninitializedNamed(rust_type(env, parameter.typ).unwrap())
+        } else {
+            use library::Type::*;
+            let type_ = env.library.type_(parameter.typ);
+            match *type_ {
+                Fundamental(fund)
+                    if fund == library::Fundamental::Utf8
+                        || fund == library::Fundamental::Filename =>
+                        {
+                            if parameter.transfer == library::Transfer::Full {
+                                NullMutPtr
+                            } else {
+                                NullPtr
+                            }
+                        }
+                _ => NullMutPtr,
+            }
+        },
+        _ => Uninitialized,
+    }
+}
+
+fn mode_to_string(mode: OutMemMode) -> String {
+    use self::OutMemMode::*;
+    match mode {
+        Uninitialized => "mem::uninitialized()".to_string(),
+        NullMutPtr => "ptr::null_mut()".to_string(),
+        UninitializedNamed(ref name) => format!("{}::uninitialized()", name),
+        NullPtr => "ptr::null()".to_string(),
+    }
+}
+
+fn crate_name(name: &str, env: &Env) -> String {
+    let id = env.library.find_namespace(name).expect("namespace from crate name");
+    let namespace = env.library.namespace(id);
+    let name = nameutil::crate_name(&namespace.name);
+    if id == namespaces::MAIN {
+        return "ffi".to_string();
+    }
+    format!("{}_ffi", name)
```

Comment:
`Namespace.ffi_crate_name` is not same?

Reply:
`ffi_crate_name` is only available on the other `Namespace` struct.